crates/ruff_python_formatter/src/expression/mod.rs

```diff
@@ -64,10 +104,75 @@ impl FormatRule<Expr, PyFormatContext<'_>> for FormatExpr {
             Expr::List(expr) => expr.format().fmt(f),
             Expr::Tuple(expr) => expr.format().fmt(f),
             Expr::Slice(expr) => expr.format().fmt(f),
-        }
+        });
+
+        let parenthesize = !self.parentheses.is_if_breaks()
+            && is_expression_parenthesized(item, f.context().contents());
+
+        let saved_level = f.context().node_level();
+        f.context_mut().set_node_level(NodeLevel::Expression);
+
+        let result = if parenthesize {
+            // Optional or Preserve and expression has parentheses in source code.
+            write!(
+                f,
+                [group(&format_args![
+                    text("("),
+                    soft_block_indent(&format_expr),
+                    text(")")
+                ])]
+            )
+        } else if !self.parentheses.is_preserve() && !has_own_parentheses(item) {
+            // Optional or IfBreaks: Add parentheses if the expression doesn't fit on a line (and doesn't have its own parentheses).
+            write!(
+                f,
+                [group(&format_args![
+                    if_group_breaks(&text("(")),
+                    soft_block_indent(&format_expr),
+                    if_group_breaks(&text(")"))
+                ])]
+            )
+        } else {
+            // Either `Preserve` and expression had no parentheses in the source code OR
+            // Expression has its own parentheses and doesn't require parenthesizing.
+            Format::fmt(&format_expr, f)
+        };
+
+        f.context_mut().set_node_level(saved_level);
+
+        result
     }
 }
 
+fn is_expression_parenthesized(expr: &Expr, contents: &str) -> bool {
+    // Search backwards to avoid ambiguity with `(a, )` and because it's faster
+    matches!(
```

Comment:
This is unfortunate. We now have to inspect the source code for every expression that we format but I haven't been able to come up with a better approach. Let's see what the benchmark says to this.

Reply:
What are the alternatives? In the initial prototype, I think I handled these similarly to comments -- so, for every parenthesis, I associated it with the child it contains.

Reply:
I considered extracting the tokenized ranges from the token stream and either:

* Store them in a vec and perform a binary search: I don't think this is faster because it trades the few character iterations with a binary search for every expression (except maybe if there are None). 
* Storing the `parenthesized` in a `Trivia` struct and look it up by node. Again, I don't think this would be faster because we trade the iterating over a few characters vs performing a hash map lookup (while fast, iterating over 1- characters is super fast too)